qiskit_addon_utils/exp_vals/expectation_values.py

```diff
@@ -0,0 +1,255 @@
+# This code is a Qiskit project.
+#
+# (C) Copyright IBM 2025.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Utility functions for calculating expectation values of observables."""
+
+from __future__ import annotations
+
+import numpy as np
+from qiskit.primitives import BitArray
+from qiskit.quantum_info import Pauli, SparseObservable, SparsePauliOp
+from qiskit.primitives.containers.observables_array import ObservablesArray
+
+def expectation_values(
+    bool_array: np.ndarray[np._bool],
+    basis_dict: dict[Pauli, list[SparsePauliOp | None]],
+    meas_basis_axis: int,
+    avg_axis: int | tuple[int] | None = None,
+    meas_flips: np.ndarray[np._bool] | None = None,
+    pec_signs: np.ndarray[np._bool] | None = None,
+    postselect_mask: np.ndarray[np._bool] | None = None,
+):
+    """Computes expectation values from boolean data.
+
+    Uses data in `bool_array`, acquired with measurement bases as ordered in keys of `basis_dict`, to compute observables encoded in values of `basis_dict`.
+
+    Optionally allows averaging over additional axes of `bool_array`, as when twirling.
+
+    Optionally supports measurement twirling, PEC, and postselection.
+
+    Args:
+        bool_array: Boolean array, presumably representing data from measured qubits.
+            The last two axes are the number of shots and number of classical bits, respectively.
+            At least one more axis must be defined indicating the measurement bases, at position `meas_basis_axis` and of length `len(basis_dict)`.
```

Comment:
Just to make sure I understand correctly: if I have a 3d array of shape `(4, 1024, 32)` that would mean I have 32 classical bits, 1024 shots and 4 different measurement bases and `meas_basis_axis` would need to be `0`, correct?
This also means that we need an explicit `meas_basis_axis` in every case, even when we have only a single measurement basis, e.g. `(1, 1024, 32)`.

Reply:
Yes, that's all correct.

Yeah, it would be better to allow `meas_basis_axis=None` in the case of 1 measurement basis... I'd defer that for now as a "nice to have" once the rest of this is working. Hopefully it's just calling reshape once at the start to add the axis, and again at the end to remove the axis before returning.